code/obj/item/device/transfer_valve.dm

```diff
@@ -138,64 +138,103 @@ TYPEINFO(/obj/item/device/transfer_valve)
+											      "maxPressure"=TANK_FRAGMENT_PRESSURE) \
+										   : list("name"=null, "num"=2, "pressure"=null, "maxPressure"=null)
+		return list(
+			"tank_one" = tank_one_data,
+			"tank_two" = tank_two_data,
+			"device" = "[src.attached_device]",
+			"opened" = src.valve_open,
+		)
+
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if (!ui)
+			ui = new(user, src, "TTV", src.name)
+			ui.open()
+
+	ui_act(action, params)
 		..()
-		if (isghostdrone(usr))
+		if (isghostdrone(usr) || usr.stat || usr.restrained())
 			return
 		if (usr.get_gang())
 			boutput(usr, SPAN_ALERT("You think working with explosives would bring a lot of much heat onto your gang to mess with this. But you do it anyway."))
-		if (usr.stat|| usr.restrained())
-			return
-		if (src.loc == usr)
-			if(href_list["tankone"])
-				tank_one.set_loc(get_turf(src))
-				tank_one = null
-				UpdateIcon()
-				if(src.equipped_in_slot)
-					var/mob/wearer = src.loc
-					wearer.update_clothing()
-			if(href_list["tanktwo"])
-				tank_two.set_loc(get_turf(src))
-				tank_two = null
-				UpdateIcon()
-				if(src.equipped_in_slot)
-					var/mob/wearer = src.loc
-					wearer.update_clothing()
-			if(href_list["open"])
-				if (valve_open)
-					var/turf/bombturf = get_turf(src)
-					logTheThing(LOG_BOMBING, usr, "closed the valve on a TTV tank transfer valve at [log_loc(bombturf)].")
-					message_admins("[key_name(usr)] closed the valve on a TTV tank transfer valve at [log_loc(bombturf)].")
-				else
-					var/turf/bombturf = get_turf(src)
-					logTheThing(LOG_BOMBING, usr, "opened the valve on a TTV tank transfer valve at [log_loc(bombturf)].")
-					message_admins("[key_name(usr)] opened the valve on a TTV tank transfer valve at [log_loc(bombturf)].")
+		switch(action)
+			if ("add_item")
+				if (ismob(usr))
+					if (params["tank"])
+						src.attach_tank(usr, params["tank"])
+					else
+						var/mob/M = usr
+						src.Attackby(M.equipped(), M)
+			if ("remove_tank_one")
+				src.remove_tank(tank_one)
+			if ("remove_tank_two")
+				src.remove_tank(tank_two)
+			if ("toggle_valve")
+				var/openorclose = (src.valve_open) ? "opened" : "closed"
+				var/turf/bombturf = get_turf(src)
+				logTheThing(LOG_BOMBING, usr, "[openorclose] the valve on a TTV tank transfer valve at [log_loc(bombturf)].")
+				message_admins("[key_name(usr)] [openorclose] the valve on a TTV tank transfer valve at [log_loc(bombturf)].")
 				toggle_valve()
-			if(href_list["rem_device"])
-				attached_device.set_loc(get_turf(src))
-				attached_device.master = null
-				attached_device = null
+			if ("remove_device")
+				src.attached_device.set_loc(get_turf(src))
+				src.attached_device.master = null
+				src.attached_device = null
 				UpdateIcon()
-			if(href_list["device"])
+			if ("interact_device")
 				attached_device.attack_self(usr)
-			if(href_list["straps"])
+			if ("remove_straps")
```

Comment:
This doesn't seem to be called anywhere? Would make more sense to be on attack with a snipping tool anyway.

Reply:
WOOPS and yea, forgot to add the straps bit to the GUI. But will be easier to just be done by a cutting tool

Reply:
added snipping tool handling to attacky